netbsd/hid.c

```diff
@@ -0,0 +1,1121 @@
+		 * out of the report descriptor.
+		 */
+
+		while (!get_next_hid_usage(ucrd->ucrd_data, ucrd->ucrd_size, &pos, &page, &usage)) {
+			/* Create new record for additional usage pairs */
+			struct hid_device_info *node = (struct hid_device_info *) calloc(1, sizeof(struct hid_device_info));
+
+			if (!node)
+				continue;
+
+			/* Update fields */
+			node->path = (end->path) ? strdup(end->path) : NULL;
+			node->vendor_id = end->vendor_id;
+			node->product_id = end->product_id;
+			node->serial_number = (end->serial_number) ? wcsdup(end->serial_number) : NULL;
+			node->release_number = end->release_number;
+			node->manufacturer_string = (end->manufacturer_string) ? wcsdup(end->manufacturer_string) : NULL;
+			node->product_string = (end->product_string) ? wcsdup(end->product_string) : NULL;
+			node->usage_page = page;
+			node->usage = usage;
+			node->interface_number = end->interface_number;
+			node->next = NULL;
+			node->bus_type = end->bus_type;
+
+			/* Insert node */
+			end->next = node;
+			end = node;
+		}
+	}
+
+	return root;
+}
+
+static int is_usb_controller(const char *s)
+{
+	return (!strncmp(s, "usb", 3) && isdigit((int) s[3]));
+}
+
+static int is_uhid_parent_device(const char *s)
+{
+	return (!strncmp(s, "uhidev", 6) && isdigit((int) s[6]));
+}
+
+static int is_uhid_device(const char *s)
+{
+	return (!strncmp(s, "uhid", 4) && isdigit((int) s[4]));
+}
+
+static void walk_device_tree(int drvctl, const char *dev, int depth, char arr[static 256][16], size_t *len, int (*cmp) (const char *))
+{
+	int res;
+	char childname[256][16];
+	struct devlistargs dla;
+
+	if (depth && (!dev || !*dev))
+		return;
+
+	if (cmp(dev) && *len < 256)
+		strlcpy(arr[(*len)++], dev, sizeof(*arr));
+
+	strlcpy(dla.l_devname, dev, sizeof(dla.l_devname));
+	dla.l_childname = childname;
+	dla.l_children = 256;
+
+	res = ioctl(drvctl, DRVLISTDEV, &dla);
+	if (res == -1)
+		return;
+
+	if (dla.l_children > 256)
+		return;
+
+	for (size_t i = 0; i < dla.l_children; i++)
+		walk_device_tree(drvctl, dla.l_childname[i], depth + 1, arr, len, cmp);
+}
+
+static void enumerate_usb_devices(int bus, uint8_t addr, enumerate_devices_callback func, void *data)
+{
+	int res;
+	struct usb_device_info udi;
+
+	udi.udi_addr = addr;
+
+	res = ioctl(bus, USB_DEVICEINFO, &udi);
+	if (res == -1)
+		return;
+
+	for (int port = 0; port < udi.udi_nports; port++) {
+		addr = udi.udi_ports[port];
+		if (addr >= USB_MAX_DEVICES)
+			continue;
+
+		enumerate_usb_devices(bus, addr, func, data);
+	}
+
+	func(&udi, data);
+}
+
+static void hid_enumerate_callback(const struct usb_device_info *udi, void *data)
+{
+	struct hid_enumerate_data *hed;
+
+	hed = (struct hid_enumerate_data *) data;
+
+	if (hed->vendor_id != 0 && hed->vendor_id != udi->udi_vendorNo)
+		return;
+
+	if (hed->product_id != 0 && hed->product_id != udi->udi_productNo)
+		return;
+
+	for (size_t i = 0; i < USB_MAX_DEVNAMES; i++) {
+		const char *parent_dev;
+		char arr[256][16];
```

Comment:
is this `256` is same as above or not?
(use same or different named constant accordingly)

Reply:
It is identical to the others. I am reusing the device tree walker function because the system call approach is the same, it just starts with a specific parent device instead of the empty string (which is the root of the device tree here).